ucp/_libs/core.pyx

```diff
@@ -313,25 +326,42 @@ cdef class ApplicationContext:
                           "ctrl_tag": hash(uuid.uuid4())}
         cdef Tags[::1] tags_mv = <Tags[:1:1]>(&tags)
 
+        # NB: we add one to the send tags on the client and we
+        #     add one to the recv tags on the server
         ep = _Endpoint(
-            PyLong_FromVoidPtr(<void*> ucp_ep),
-            PyLong_FromVoidPtr(<void*> self.worker),
-            self.config,
-            tags.msg_tag,
-            tags.ctrl_tag,
+            ucp_endpoint=PyLong_FromVoidPtr(<void*> ucp_ep),
+            ucp_worker=PyLong_FromVoidPtr(<void*> self.worker),
+            config=self.config,
+            msg_tag_send=tags.msg_tag+1,
+            msg_tag_recv=tags.msg_tag,
+            ctrl_tag_send=tags.ctrl_tag+1,
```

Comment:
What are the `+1`s here for?

Reply:
This is what differentiate the send/recv tags :)
I have added some more comments to make more clear